File: Reverse Linked List/JavaScript/solution1.js
// Not the fastest but take advantage of some JS features
// Load the list  into an array then reverse the array and set the next values

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
 var reverseList = function(head) {
    var listArr = [];
    var node = head;
    // load into an array
    while(node != null){
        listArr.push(node);
        node = node.next;
    }
    // reverse the array
    listArr.reverse();
    // iterate over the array
    for(var i = 0; i < listArr.length; i++){
        // if it's the last item in the array then it points to null else it points to the next index
        listArr[i].next = (i+1 < listArr.length) ? listArr[i+1] : null;
    }
    // return either the first node or empty depending
    return listArr[0] ? listArr[0] : [];
};

File: Reverse Linked List/Go/solution1.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

// Recursive solution
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func reverseList(head *ListNode) *ListNode {
	// head is nil or next is nil return the head. Exit condition for recursion.
	if head == nil || head.Next == nil {
		return head
	}
	// recursively call reverseList until the exit condition is hit. The returned node is the new node and last node in list
	nh := reverseList(head.Next)
	// The node after the current node should point to the current node
	head.Next.Next = head
	// The current node will point to nothing until set in the previously called recursive function
	head.Next = nil

	// return the new head
	return nh
}
